src/category_theory/abelian/projective.lean

```diff
@@ -0,0 +1,200 @@
+-/
+class projective (P : C) : Prop :=
+(factors : ∀ {E X : C} (f : P ⟶ X) (e : E ⟶ X) [epi e], ∃ f', f' ≫ e = f)
+
+section
+
+/--
+A projective presentation of an object `X` consists of an epimorphism `f : P ⟶ X`
+from some projective object `P`.
+-/
+@[nolint has_inhabited_instance]
+structure projective_presentation (X : C) :=
+(P : C)
+[projective : projective P]
+(f : P ⟶ X)
+(epi : epi f)
+
+variables (C)
+
+/-- A category "has enough projectives" if for every object `X` there is a projective object `P` and
+    an epimorphism `P ↠ X`. -/
+class enough_projectives : Prop :=
+(presentation : ∀ (X : C), nonempty (projective_presentation X))
+
+end
+
+namespace projective
+
+lemma of_iso {P Q : C} (i : P ≅ Q) (hP : projective P) : projective Q :=
+begin
+  fsplit,
+  introsI E X f e e_epi,
+  obtain ⟨f', hf'⟩ := projective.factors (i.hom ≫ f) e,
+  exact ⟨i.inv ≫ f', by simp [hf']⟩
+end
+
+lemma iso_iff {P Q : C} (i : P ≅ Q) : projective P ↔ projective Q :=
+⟨of_iso i, of_iso i.symm⟩
+
+section enough_projectives
+variables [enough_projectives C]
+
+/--
+`projective.over X` provides an arbitrarily chosen projective object equipped with
+an epimorphism `projective.π : projective.over X ⟶ X`.
+-/
+def over (X : C) : C :=
+(enough_projectives.presentation X).some.P
+
+instance projective_over (X : C) : projective (over X) :=
+(enough_projectives.presentation X).some.projective
+
+/--
+The epimorphism `projective.π : projective.over X ⟶ X`
+from the arbitrarily chosen projective object over `X`.
+-/
+def π (X : C) : over X ⟶ X :=
+(enough_projectives.presentation X).some.f
+
+instance π_epi (X : C) : epi (π X) :=
+(enough_projectives.presentation X).some.epi
+
+section
+variables [has_zero_morphisms C] {X Y : C} (f : X ⟶ Y) [has_kernel f]
+
+/-- When `C` has enough projectives, the object `projective.left f` is
+the arbitrarily chosen projective object over `kernel f`. -/
+@[derive projective]
+def left : C := over (kernel f)
+
+/-- When `C` has enough projectives,
+`projective.d f : projective.left f ⟶ X` is the composition
+`π (kernel f) ≫ kernel.ι f`.
+-/
+abbreviation d : left f ⟶ X :=
+π (kernel f) ≫ kernel.ι f
+
+end
+
+section
+variables [abelian C]
+
+/--
+When `C` is abelian, `projective.d f` and `f` are exact.
+-/
+lemma exact_d_f {X Y : C} (f : X ⟶ Y) : exact (d f) f :=
+(abelian.exact_iff _ _).2 $
+  ⟨by simp, zero_of_epi_comp (π _) $ by rw [←category.assoc, cokernel.condition]⟩
+
+end
+
+section
+variables [has_zero_morphisms C] [has_equalizers C] [has_images C]
+
+/--
+A `projective.resolution Z` is an exact sequence `... ⟶ X 2 ⟶ X 1 ⟶ X 0`,
+where `X 0 = Z`, and the `X (n+1)` are projective.
+
+For simplicity we define a structure here independent of our implementation of chain complexes,
+and will later connect that up.
+-/
+@[nolint has_inhabited_instance]
+structure resolution (Z : C) :=
+(X : ℕ → C)
+(d : Π n, X (n+1) ⟶ X n)
+(zero : X 0 = Z)
```

Comment:
Can this equality of objects become a problem at some point? It would be much nicer if there was some way to assert that `X 0` and `Z` were defeq.

Reply:
Perhaps we should just relax this to an isomorphism.